mmaction/core/multi_grid_utils.py

```diff
@@ -0,0 +1,252 @@
+import numpy as np
+import torch.nn as nn
+from mmcv.runner import HOOKS, Hook, LrUpdaterHook
+from mmcv.runner.hooks.lr_updater import StepLrUpdaterHook
+from torch.nn.modules.utils import _ntuple
+
+from mmaction.datasets.builder import build_dataloader
+from mmaction.datasets.pipelines.augmentations import Resize
+from mmaction.datasets.pipelines.loading import SampleFrames
+from mmaction.models.common import SubBatchBN3d
+from ..utils import get_root_logger
+
+
+def modify_num_splits(module, num_splits):
+    """Recursively modify the number of splits of subbn3ds in module.
+
+    Args:
+        module (nn.Module): The module to be modified.
+        num_splits (int): The targeted number of splits.
+
+    Returns:
+        int: The number of subbn3d modules modified.
+    """
+    count = 0
+    for child in module.children():
+        if isinstance(child, SubBatchBN3d):
+            new_split_bn = nn.BatchNorm3d(child.num_features *
+                                          num_splits).cuda()
+            child.num_splits = num_splits
+            child.split_bn = new_split_bn
+            count += 1
+        else:
+            count += modify_num_splits(child, num_splits)
+    return count
+
+
+@HOOKS.register_module()
+class RelativeStepLrUpdaterHook(LrUpdaterHook):
+    """RelativeStepLrUpdaterHook.
+
+    Args:
+        ori_step (list[int]): Same as that of mmcv.
+        new_step (list[int]): The real steps caused by multi-grid
+            training.
+        step_lr_ratio (list[int]): The extra factor to be multiplied
+            at each new step.
+        gamma (float): Same as that of mmcv.
+        **kwargs (dict): Same as that of mmcv.
+    """
+
+    def __init__(self, runner, step, lrs, **kwargs):
+        super().__init__(**kwargs)
+        assert len(step) == (len(lrs) - 1)
+        self.step = step
+        self.lrs = lrs
+        super().before_run(runner)
```

Comment:
Is this line necessary? [It](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/epoch_based_runner.py#L103) seems the `before_run` will be automatically call.

Reply:
Yeah, it is because necessary attributes such as `self.base_lr` is initiated in `before_run` of a LrHook, but when I swap one with the new one. The new one will not be initialized.

Reply:
Another solution is just move the contents in `before_run` to `__init__`, it will be better to understand.

Reply:
Oh I know, it may be casued by the order of the hook. Since the original `StepLrHook` called `before_run` first, and the new one `LrHook` will not call `before_run`, resulting in the no initialization. So, just `super().before_run(runner)` is OK.